# Week 1 Demo: Python as a Calculator, Variables, Data Types, Casting, and Formatted Printing

Welcome to your first hands-on Python notebook! This notebook mixes short explanations, **demo cells** (already run, so you can see the output), and **"Your Turn" exercise cells** (blank, for you to fill in and run yourself).

Follow along top to bottom, run every demo cell yourself as you go (even though the output is already shown), and complete each "Your Turn" exercise before moving to the next section.

*(Style and structure inspired by UC Berkeley's Data 6 course notes: [data6.org/notes](https://data6.org/notes/))*


## 1. Python as a Calculator

Before Python does anything fancy, it's just a very capable calculator. You can type an arithmetic expression directly into a code cell and run it.

| Operator | Meaning | Example |
|---|---|---|
| `+` | addition | `2 + 3` |
| `-` | subtraction | `5 - 1` |
| `*` | multiplication | `4 * 3` |
| `/` | division (always returns a float) | `7 / 2` |
| `//` | floor division (rounds down to an integer) | `7 // 2` |
| `%` | modulo (remainder after division) | `7 % 2` |
| `**` | exponent (power) | `2 ** 3` |

Python follows the standard order of operations (PEMDAS) — parentheses first, then exponents, then multiplication/division, then addition/subtraction.

In [1]:
2 + 3

Notice that Jupyter automatically displays the result of the **last line** in a cell, even without `print()`. This only works for the last line — if you want to see multiple results, you need `print()`.

In [2]:
print(7 / 2)    # true division -- always a float
print(7 // 2)   # floor division -- rounds down
print(7 % 2)    # modulo -- the remainder
print(2 ** 3)   # exponent -- 2 to the power of 3

3.5
3
1
8


In [3]:
# Order of operations matters!
print(2 + 3 * 4)      # multiplication happens first: 2 + 12 = 14
print((2 + 3) * 4)    # parentheses override that: 5 * 4 = 20

14
20


### Your Turn

In the cell below:
1. Compute how many **full weeks** are in 100 days (hint: floor division)
2. Compute how many **leftover days** remain after those full weeks (hint: modulo)
3. Print both results, clearly labeled


In [ ]:
# Your Turn: full weeks and leftover days in 100 days
days = 100

# TODO: compute full weeks using floor division
full_weeks = ...

# TODO: compute leftover days using modulo
leftover_days = ...

# TODO: print both, labeled


## 2. Variables

A **variable** is a name that stores a value so you can reuse it later, instead of retyping the value every time. In Python, you create a variable with a single equals sign `=` (this is *assignment*, not mathematical equality!).

**Naming rules:**
- Must start with a letter or underscore (not a number)
- Can contain letters, numbers, and underscores
- Case-sensitive (`age` and `Age` are different variables)
- Cannot be a Python reserved word (like `for`, `if`, `print`)

**Naming convention:** use `snake_case` (lowercase with underscores) for variable names in Python, e.g. `student_count`, not `StudentCount` or `studentCount`.


In [4]:
collector_area = 2.4       # square meters
efficiency = 0.55
irradiance = 700           # watts per square meter

power_absorbed = collector_area * efficiency * irradiance
print(power_absorbed)

924.0


A variable can be **reassigned** at any time — Python doesn't lock in the original value or type:

In [5]:
x = 5
print(x)
x = x + 1      # take the current value of x, add 1, store it back in x
print(x)
x = "now I'm a string"   # variables can even change type
print(x)

5
6
now I'm a string


### Your Turn

A small drone has a battery rated at `battery_mAh = 2200` and `battery_V = 11.1`. Battery energy in watt-hours is `(mAh / 1000) * V`.

1. Create variables `battery_mAh` and `battery_V` with the values above
2. Compute the battery energy in watt-hours and store it in a variable called `battery_Wh`
3. Print the result, labeled


In [ ]:
# Your Turn: battery energy calculation
battery_mAh = ...
battery_V = ...

# TODO: compute battery_Wh

# TODO: print, labeled


## 3. Data Types

Every value in Python has a **type**. The four most common basic types you'll use constantly:

| Type | Name | Example | Notes |
|---|---|---|---|
| `int` | integer | `42`, `-7` | whole numbers, no decimal point |
| `float` | floating-point number | `3.14`, `-0.5`, `2.0` | has a decimal point |
| `str` | string | `"hello"`, `'x'` | text, in single or double quotes |
| `bool` | boolean | `True`, `False` | exactly one of two values (capitalized!) |

You can check any value's type with the built-in `type()` function.


In [6]:
print(type(42))
print(type(3.14))
print(type("hello"))
print(type(True))

<class 'int'>
<class 'float'>
<class 'str'>
<class 'bool'>


A common surprise: `5` and `5.0` are *not* the same type, even though they represent the same number:

In [7]:
print(type(5))
print(type(5.0))
print(5 == 5.0)   # they ARE considered equal in value, just not the same type

<class 'int'>
<class 'float'>
True


### Your Turn

For each value below, predict its type, then check yourself using `type()`:
- `10`
- `10.0`
- `"10"`
- `10 == 10.0`


In [ ]:
# Your Turn: check the type of each value below
# TODO: print type(10)

# TODO: print type(10.0)

# TODO: print type("10")

# TODO: print type(10 == 10.0)


## 4. Type Casting

**Casting** means explicitly converting a value from one type to another, using the type name as a function: `int(...)`, `float(...)`, `str(...)`, `bool(...)`.

This comes up constantly when working with real data — e.g., a value read from a file or user input often arrives as a `str`, even if it "looks like" a number, and you have to cast it before doing math with it.


In [8]:
age_str = "20"
print(type(age_str))

age_int = int(age_str)    # cast string -> int
print(type(age_int))
print(age_int + 5)         # now math works

<class 'str'>
<class 'int'>
25


Casting `float` &rarr; `int` **truncates** (chops off the decimal part) rather than rounding:

In [9]:
print(int(9.99))   # truncates to 9, NOT rounded to 10
print(int(9.01))   # also 9
print(round(9.99)) # round() is the function to use if you actually want rounding

9
9
10


Not every cast is valid — casting a non-numeric string to `int` raises an error:

In [10]:
try:
    int("hello")
except ValueError as e:
    print("This failed, as expected:", e)

This failed, as expected: invalid literal for int() with base 10: 'hello'


### Your Turn

A sensor logs a temperature reading as the string `"98.6"`. 

1. Cast it to a `float` and store it as `temp_float`
2. Cast `temp_float` to an `int` and store it as `temp_int` — predict what you'll get *before* running the cell
3. Print both, labeled


In [ ]:
# Your Turn: casting a sensor reading
temp_str = "98.6"

# TODO: cast to float

# TODO: cast the float to int

# TODO: print both, labeled


## 5. Formatted Printing (Python's version of `printf`)

If you've used C, you might know `printf("%d apples", count)`. Python has an equivalent — and a more modern, easier-to-read alternative.

### Old style: `%` formatting (works like C's `printf`)


In [11]:
name = "Alex"
score = 92.5
print("Student %s scored %.1f points" % (name, score))
# %s = string, %.1f = float rounded to 1 decimal place, %d = integer

Student Alex scored 92.5 points


### Modern style: f-strings (recommended)

An **f-string** is a string with an `f` right before the opening quote. Anything inside `{}` is evaluated as Python code and inserted directly into the string. This is the standard, recommended way to format strings in modern Python (3.6+), and what you'll see used throughout this course.


In [12]:
name = "Alex"
score = 92.5
print(f"Student {name} scored {score} points")

Student Alex scored 92.5 points


f-strings also support format specifiers after a colon `:`, for controlling decimal places, padding, etc. — this is the direct equivalent of `printf`'s `%.2f`-style formatting:

In [13]:
pi = 3.14159265
print(f"Pi rounded to 2 decimal places: {pi:.2f}")
print(f"Pi rounded to 4 decimal places: {pi:.4f}")

count = 7
print(f"Count padded to 3 digits: {count:03}")   # 007

large_num = 1234567
print(f"With thousands separator: {large_num:,}")

Pi rounded to 2 decimal places: 3.14
Pi rounded to 4 decimal places: 3.1416
Count padded to 3 digits: 007
With thousands separator: 1,234,567


### Your Turn

Using an f-string, print a sentence in the form:

```
The collector absorbs 924.00 W at 55% efficiency.
```

using the variables below (the power value should show exactly 2 decimal places, and the efficiency should be shown as a whole-number percentage).


In [ ]:
# Your Turn: formatted printing with f-strings
power_absorbed = 924.0
efficiency = 0.55

# TODO: print the sentence above using an f-string,
#       with power_absorbed shown to 2 decimal places
#       and efficiency shown as a whole-number percentage (hint: multiply by 100)


## 6. Complex Number Arithmetic

Python has built-in support for complex numbers — no import needed for basic arithmetic. A complex number is written with a `j` (not `i`, as in math) directly after the imaginary part's coefficient, e.g. `3 + 4j`.

| Operation | Syntax | Notes |
|---|---|---|
| Create | `3 + 4j` or `complex(3, 4)` | both create the same value |
| Real part | `z.real` | always returns a `float` |
| Imaginary part | `z.imag` | always returns a `float` |
| Conjugate | `z.conjugate()` | flips the sign of the imaginary part |
| Magnitude | `abs(z)` | $\sqrt{\text{real}^2 + \text{imag}^2}$ |
| Add, subtract, multiply, divide | `+`, `-`, `*`, `/` | work exactly like normal arithmetic |


In [14]:
z1 = 3 + 4j
z2 = 1 - 2j

print("z1 =", z1)
print("z2 =", z2)
print("type(z1):", type(z1))

z1 = (3+4j)
z2 = (1-2j)
type(z1): <class 'complex'>


In [15]:
print("z1 + z2 =", z1 + z2)
print("z1 - z2 =", z1 - z2)
print("z1 * z2 =", z1 * z2)
print("z1 / z2 =", z1 / z2)

z1 + z2 = (4+2j)
z1 - z2 = (2+6j)
z1 * z2 = (11-2j)
z1 / z2 = (-1+2j)


In [16]:
print("Real part of z1:", z1.real)
print("Imaginary part of z1:", z1.imag)
print("Conjugate of z1:", z1.conjugate())
print("Magnitude of z1 (|z1|):", abs(z1))   # sqrt(3^2 + 4^2) = 5.0

Real part of z1: 3.0
Imaginary part of z1: 4.0
Conjugate of z1: (3-4j)
Magnitude of z1 (|z1|): 5.0


A common gotcha: writing `3 + 4i` (using `i` like in math class, with no `*`) is actually a **SyntaxError** in Python — it must be `j`, written directly against the number with no space:

In [17]:
try:
    exec("bad = 3 + 4i")
except SyntaxError as e:
    print("This failed, as expected:", e)

This failed, as expected: invalid decimal literal (<string>, line 1)


### Your Turn

Two complex impedances in an AC circuit are $Z_1 = 4 + 3j$ ohms and $Z_2 = 2 - 5j$ ohms.

1. Compute their sum `Z_total = Z1 + Z2` (impedances in series simply add)
2. Print the real and imaginary parts of `Z_total` separately, labeled
3. Print the magnitude of `Z_total` using `abs()`


In [ ]:
# Your Turn: complex impedance in series
Z1 = 4 + 3j
Z2 = 2 - 5j

# TODO: compute Z_total

# TODO: print the real and imaginary parts of Z_total, labeled

# TODO: print the magnitude of Z_total


## 7. Floating-Point Numbers, the `math` Library, and Why `==` Can Lie to You

You've already seen that `float` is one of Python's basic types. This section goes deeper: floats aren't stored with perfect precision, which causes some genuinely surprising behavior you need to know about before you write code that compares numbers.

### More Floating-Point Arithmetic Examples

Floats support the same operators as integers, but division, roots, and repeating decimals often produce long, imprecise-looking results:

In [18]:
print(1 / 3)
print(2 / 3)
print(10 / 4)
print(2 ** 0.5)      # square root of 2, using the exponent operator
print(3.3 - 3.1)     # you might expect exactly 0.2 -- see what you actually get

0.3333333333333333
0.6666666666666666
2.5
1.4142135623730951
0.19999999999999973


That last one is the first hint of the problem: `3.3 - 3.1` doesn't print as a clean `0.2`. Computers store floats in binary, and most decimal fractions (like 0.1, 0.2, 0.3) **cannot be represented exactly** in binary — the same way $\frac{1}{3}$ can't be written exactly as a finite decimal. Python is doing the math correctly; it's just displaying the tiny leftover imprecision.

### The Classic Floating-Point Comparison Trap

This is one of the most common bugs in scientific and engineering code: comparing two floats with `==` when you actually meant "close enough."

In [19]:
print(0.1 + 0.2)
print(0.1 + 0.2 == 0.3)

0.30000000000000004
False


That printed `False`! Let's see why, by looking at the *actual* stored value with more decimal places using `repr()`:

In [20]:
print(repr(0.1 + 0.2))
print(repr(0.3))

0.30000000000000004
0.3


`0.1 + 0.2` is actually stored as `0.30000000000000004` — off from `0.3` by about $4 \times 10^{-17}$. That's a tiny error, but `==` checks for *exact* equality, so even a difference this small makes the comparison `False`. This is not a Python bug — it happens in essentially every programming language, because it's a consequence of binary floating-point representation (IEEE 754), not a language design choice.

**A second classic example**, familiar from earlier in the course: adding a small number to a very large one can lose it entirely.

In [21]:
a = 1e16
print((a + 1) - a)   # you'd expect 1.0 -- but the '+1' doesn't fit in the available precision

0.0


### The Fix: `math.isclose()` Instead of `==`

The correct way to compare floats is to check whether they're *close enough*, not exactly equal. Python's built-in `math` library provides `math.isclose()` for exactly this purpose.

In [22]:
import math

print(math.isclose(0.1 + 0.2, 0.3))              # True -- this is the fix
print(math.isclose(1.0000001, 1.0000002))          # False by default -- very tight tolerance
print(math.isclose(1.0000001, 1.0000002, abs_tol=1e-5))  # True -- loosened tolerance

True
False
True


`math.isclose()` has two tolerance settings you can control: `rel_tol` (relative tolerance, a percentage of the larger number) and `abs_tol` (absolute tolerance, a fixed cushion — important when comparing numbers near zero, where a purely relative tolerance breaks down). For most engineering calculations, setting an explicit `abs_tol` appropriate to your problem is the safest habit.

### The `math` Library: Common Functions

Python's built-in `math` library provides functions and constants beyond basic arithmetic. You must `import math` before using it.

In [23]:
import math

print(math.sqrt(2))          # square root
print(math.pi)                # the constant pi
print(math.e)                 # Euler's number
print(math.floor(4.7))        # rounds down to the nearest integer
print(math.ceil(4.2))         # rounds up to the nearest integer
print(math.factorial(5))      # 5! = 5*4*3*2*1

1.4142135623730951
3.141592653589793
2.718281828459045
4
5
120


Note: `math.sqrt()` only works on real, non-negative numbers — for complex numbers (Section 6 above), you'd use the separate `cmath` library instead, or simply use `**0.5`, which works for both.

In [24]:
try:
    math.sqrt(-4)
except ValueError as e:
    print("This failed, as expected:", e)

print((-4) ** 0.5)   # this works and returns a complex result instead

This failed, as expected: math domain error
(1.2246467991473532e-16+2j)


### Your Turn

1. Use `math.sqrt()` to compute the hypotenuse of a right triangle with legs 3 and 4 (hint: $\sqrt{3^2 + 4^2}$)
2. Separately, add `0.1` three times (`0.1 + 0.1 + 0.1`) and compare the result to `0.3` using `==` — what do you get?
3. Now compare the same two values using `math.isclose()` instead — confirm it correctly reports them as equal


In [ ]:
# Your Turn
import math

# TODO 1: compute the hypotenuse of a 3-4-5 triangle using math.sqrt()

# TODO 2: compare 0.1 + 0.1 + 0.1 to 0.3 using ==, and print the result

# TODO 3: compare the same two values using math.isclose(), and print the result


## 8. Getting Input From the User

So far, every value in this notebook has been hard-coded directly into your script. Python's built-in `input()` function lets a running program pause and ask the person using it to type something in.

**Important:** `input()` **always** returns a `str`, no matter what the user types — even if they type `"42"`, you get back the *string* `"42"`, not the *number* `42`. If you want to do math with it, you must cast it first (Section 4 above).

*(The example cells below simulate a user typing a response, so you can see what the output looks like without needing to actually type anything here. When you run these cells yourself in a live notebook, Python will pause and show you a real input box — try it!)*

### Example 1: String Input

In [25]:
user_name = input("What is your name? ")
print(f"Hello, {user_name}! Welcome to ENGR 50.")

What is your name? Hello, Sam! Welcome to ENGR 50.


### Example 2: Numeric Input (remember to cast!)

In [26]:
age_str = input("How old are you? ")   # this is a STRING, even though it looks like a number
print("Type of what input() returned:", type(age_str))

age = int(age_str)                       # cast to int before doing math
years_until_30 = 30 - age
print(f"You have {years_until_30} years until you turn 30.")

How old are you? Type of what input() returned: <class 'str'>
You have 9 years until you turn 30.


### Example 3: Combining String and Numeric Input

In [27]:
name = input("What is your name? ")
temp_str = input("What is the current temperature in Fahrenheit? ")

temp_F = float(temp_str)
temp_C = (temp_F - 32) * 5/9

print(f"Hi {name}, {temp_F} F is {temp_C:.1f} C.")

What is your name? What is the current temperature in Fahrenheit? Hi Sam, 77.4 F is 25.2 C.


### Your Turn

Write code that:
1. Asks the user to input a distance in **meters** (as a string, via `input()`)
2. Casts it to a `float`
3. Converts it to **feet** (1 meter = 3.28084 feet)
4. Prints the result in a labeled sentence using an f-string

To test your code in this notebook, run the cell and actually type a number when prompted.

In [ ]:
# Your Turn: convert user-entered meters to feet
# TODO: use input() to get a distance in meters

# TODO: cast to float

# TODO: convert to feet (1 m = 3.28084 ft)

# TODO: print the result, labeled


## 9. Simple String "Math": `+` and `*`

Strings support two of the same operators you've been using on numbers — but they do something different: `+` **concatenates** (joins) strings together, and `*` **repeats** a string a given number of times.

*(This is just a preview — we'll explore strings in much more detail, including indexing, slicing, and string methods, in a later notebook.)*

In [28]:
# + concatenates strings
first = "Py"
second = "thon"
print(first + second)

greeting = "Hello" + ", " + "world" + "!"
print(greeting)

Python
Hello, world!


In [29]:
# * repeats a string
print("ab" * 3)
print("=" * 20)      # a quick way to print a divider line
print("Ha" * 4 + "!")

ababab
HaHaHaHa!


Note: `+` requires **both sides to be strings** — you can't directly add a string and a number, even if the number is inside a variable. This is a very common early error:

In [30]:
try:
    print("Age: " + 21)
except TypeError as e:
    print("This failed, as expected:", e)

# The fix: cast the number to a string first
print("Age: " + str(21))

This failed, as expected: can only concatenate str (not "int") to str
Age: 21


### Your Turn

Using only `+` and `*` (no f-strings this time), build and print a small "banner" like this:

```
****************
*  ENGR 50  *
****************
```

(Exact spacing is up to you — the point is practicing string concatenation and repetition.)

In [ ]:
# Your Turn: build a banner using string + and *
# TODO: build the top/bottom border using *

# TODO: build the middle line using +

# TODO: print all three lines


## Wrap-Up Challenge

Put everything from this notebook together. A weather station reports temperature as a **string** in Fahrenheit: `"77.4"`. 

Write code that:
1. Casts the string to a `float`
2. Converts it to Celsius using $C = (F - 32) \times \frac{5}{9}$
3. Prints a labeled sentence with the Celsius value rounded to 1 decimal place, using an f-string

Example target output: `77.4°F is 25.2°C`


In [ ]:
# Wrap-Up Challenge
temp_F_str = "77.4"

# TODO: cast to float

# TODO: convert to Celsius

# TODO: print the labeled result using an f-string, rounded to 1 decimal place
